### PR TITLE
Add TLS support between Ironic and BMO

### PR DESCRIPTION
**Pull request by maelk**
This PR introduces support for TLS using a custom CA certificate in BMO and modifies the deployment file to be able to deploy Ironic, inspector and BMO with TLS enabled.

This requires https://github.com/metal3-io/ironic-image/pull/198 and https://github.com/metal3-io/ironic-inspector-image/pull/62

### CODE REVIEW

File: ironic-deployment/keepalived-tls/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: metal3
+resources:
+- ../keepalived
+
+configMapGenerator:
+- behavior: merge
+  literals:
+  - CACERT_FILE=/tls/ca/ca.crt
```
Comment (dtantsur) [inline ironic-deployment/keepalived-tls/kustomization.yaml:10]: Do you mean to use client certificate validation? As I mentioned on the ironic-image review, I'd skip it, at least for now. We don't have a way to provide client keys to the ramdisk (nor do you propose one here).

File: ironic-deployment/keepalived-tls/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  name: ironic-bmo-configmap
+
+secretGenerator:
+- name: ironic-cacert
+  files:
+  - ca.crt=ca.crt
+- name: ironic-cert
+  files:
+  - tls.crt=tls.crt
+  - tls.key=tls.key
```
Comment (dtantsur) [inline ironic-deployment/keepalived-tls/kustomization.yaml:22]: Only this is actually a secret.

File: pkg/provisioner/ironic/ironic.go [hunk 3]
```diff
@@ -84,6 +85,16 @@ func init() {
 		fmt.Fprintf(os.Stderr, "Cannot start: No IRONIC_INSPECTOR_ENDPOINT variable set")
 		os.Exit(1)
 	}
+	ironicTrustedCAFile = os.Getenv("IRONIC_CACERT_FILE")
```
Comment (dtantsur) [inline pkg/provisioner/ironic/ironic.go:88]: `IRONIC_CERT_FILE`. You don't need CACERT, at least if I understand your motivation right.

**Review by dtantsur — CHANGES_REQUESTED** [review]
I think there is a confusion around CA cert vs cert. Unless you want to establish a proper chain OR use client certificates, you don't need 3 files. You only need a private key (passed to ironic as a secret) and a certificate (used everywhere else). I suggest we limit at least the initial proposal to this simplest case to avoid confusion. I'd also prefer we generate certificates rather than make an operator do it.

File: docs/deploying.md [hunk 4]
```diff
@@ -90,6 +96,48 @@ kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
 
 where $BMOPATH points to the baremetal-operator path.
 
+### Deploying with TLS
+
+For this, you need to first copy your Ironic CA Certificate, then build the
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply docs/deploying.md:101]: Is the goal to allow a deployment script (or helm chart or whatever) to create the files?

[the remark this answers is not in the document]
Comment (maelk) [reply docs/deploying.md:101]: In our case, we want to be able to provide the CA for the ironic deployment. Ironic itself would use the certificates generated by cert-manager, but for BMO to accept them all, we need to use the CA, not a specific certificate.
@dhellmann yes, at the end it will be cert-manager creating the certificates and keys, however, it could be given the CA to do so.

File: ironic-deployment/keepalived-tls/kustomization.yaml (hunk 1 above)
Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml:10]: no, the mTLS part is not supported due to issue with PHA in go. We are going for plain TLS, but we want to be able to accept any certificate genearted by cert-manager using that CA

File: ironic-deployment/keepalived-tls/kustomization.yaml (hunk 2 above)
Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml:22]: Indeed, Ironic itself probably does not need the CA, only BMO does

File: ironic-deployment/keepalived-tls/kustomization.yaml (hunk 1 above)
Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml:10]: I'll remove the CA part from Ironic configuration. it is only needed in BMO actually, considering Ironic has the certificates.

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
Comment (maelk) [reply pkg/provisioner/ironic/ironic.go:88]: We need a CA to be able to support the cert-manager certificate rotation. However,the user could also just provide the certificate itself.

**Review by maelk — COMMENTED** [review]
Let me clarify a bit our scenario. We do not want to do client authentication, that is out of the picture, so as @dtantsur pointed out, we do not need the CA in Ironic, just the certs. However, we want to deploy ironic using cert-manager to generate and rotate the certificates used, based on a CA given by the user (or automatically generated). In that case, BMO should accept any certificate signed by that CA, and not only the current certificate, keeping in mind that there isn't any tight link between ironic and BMO anymore, they could even be deployed in different clusters. Hence, not taking the specific certificate, but the CA, allows for not having to re-deploy BMO when cert-manager rotates the ironic certificate.

**Comment by maelk** [comment]
I have now modified the ironic images. we are providing the certificate and the key for ironic and inspector, and in addition the CA to validate the certificate to both of them so that they can validate each other's certificate . we provide that certificate to BMO to validate the TLS connection with ironic and inspector. @dtantsur and @dhellmann please take a look

File: docs/deploying.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply docs/deploying.md:101]: > we need to use the CA, not a specific certificate

This is not really correct: in the simplest case you can use the certificate. I don't disagree there is a valid case for having a CA, but I would prefer that we also support a simple one-knob configuration "get_me_tls: true" where we'll just generate a pair key-certificate and pass certificate as both CERT_FILE and CACERT_FILE.

**Comment by dtantsur** [comment]
> Hence, not taking the specific certificate, but the CA, allows for not having to re-deploy BMO when cert-manager rotates the ironic certificate.

Got it, thank you for explanation.

File: ironic-deployment/keepalived-tls/kustomization.yaml (hunk 2 above)
Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml:22]: I added the behaviour of using the certificate as CA for ironic, so now we can remove this.

File: docs/deploying.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (maelk) [reply docs/deploying.md:101]: ok, this is now supported, we can just generate a certificate and a key, and pass it to both ironic and BMO and get it going by passing the same file for IRONIC_CACERT_FILE in BMO.

File: pkg/provisioner/ironic/clients/client.go [hunk 5]
```diff
@@ -26,11 +49,14 @@ func IronicClient(ironicEndpoint string, auth AuthConfig) (client *gophercloud.S
...
-	return
+	if err != nil {
+		return
+	}
+	return updateHTTPClient(client, caFile)
 }
 
 // InspectorClient creates a client for Ironic Inspector
-func InspectorClient(inspectorEndpoint string, auth AuthConfig) (client *gophercloud.ServiceClient, err error) {
+func InspectorClient(inspectorEndpoint string, auth AuthConfig, caFile string) (client *gophercloud.ServiceClient, err error) {
```
Comment (zaneb) [inline pkg/provisioner/ironic/clients/client.go:59]: Why not make this part of the AuthConfig struct?

File: pkg/provisioner/ironic/clients/client.go [hunk 6]
```diff
@@ -46,5 +72,8 @@ func InspectorClient(inspectorEndpoint string, auth AuthConfig) (client *gopherc
 	default:
 		err = fmt.Errorf("Unknown auth type %s", auth.Type)
 	}
-	return
+	if err != nil {
+		return
+	}
+	return updateHTTPClient(client, caFile)
```
Comment (zaneb) [inline pkg/provisioner/ironic/clients/client.go:78]: I really like that this is orthogonal to whether we're using basic auth or not :+1:

File: pkg/provisioner/ironic/clients/client.go (hunk 5 above)
Comment (maelk) [reply pkg/provisioner/ironic/clients/client.go:59]: good point, I will update

Comment (maelk) [reply pkg/provisioner/ironic/clients/client.go:59]: actually, it's handled quite differently, and the path to the CA is unrelated to the authConfig, so I think keeping it separated might be cleaner/clearer. What do you think ?

Comment (zaneb) [reply pkg/provisioner/ironic/clients/client.go:59]: I missed that this is just TLS, not mTLS, so it's only auth if you squint (technically we are authenticating the server!). It probably does make sense to keep it separate if:

* There are no other options likely to be added (e.g. use https but allow self-signed certs); and
* When mTLS support is added, this remains completely independent (they don't e.g. need to share a CA)

Comment (maelk) [reply pkg/provisioner/ironic/clients/client.go:59]: I'm playing with the idea of making it a TLSConfig. I'll modify the PR with that, please take a look.

**Comment by kashifest** [comment]
lgtm

**Comment by furkatgofurov7** [comment]
lgtm

File: deploy/tls/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+resources:
+- ../default
+
+configMapGenerator:
+- behavior: merge
+  literals:
+    - IRONIC_CACERT_FILE=/tls/ironic_ca.crt
+  name: ironic-bmo-configmap
+
+secretGenerator:
```
Comment (dtantsur) [inline deploy/tls/kustomization.yaml:13]: Not sure what this means, but a certificate is not a secret.

File: docs/deploying.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply docs/deploying.md:101]: Let's mention that the CA certificate is no longer required?

File: docs/deploying.md [hunk 8]
```diff
@@ -90,6 +96,48 @@ kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
...
+   cp <path-to-ca-certificate> $BMOPATH/deploy/tls/ca.crt
+   ./tools/bin/kustomize build $BMOPATH/deploy/tls | kubectl apply -f -
+```
+
+Then you need to copy all your certificates and build the kustomization for
+ironic:
+
+```sh
+   cp <path-to-ca-certificate> $BMOPATH/ironic-deployment/tls/ca.crt
+   cp <path-to-tls-certificate> $BMOPATH/ironic-deployment/tls/tls.crt
```
Comment (dtantsur) [inline docs/deploying.md:114]: I'm curious how you're planning for the certificate to be generated by a user if they don't know an IP address and we don't have host names.

Comment (maelk) [reply docs/deploying.md:114]: If they don't know the ip address, they can't really do any setup, no ? How would they configure BMO ? What they could maybe do is generate a certificate for a domain name that also covers the 127.0.0.1 IP address with the SAN, give that as PROVISIONING_IP to ironic and inspector, and configure BMO with the domain (for example from k8s service). That could work. But if the user is deploying with the keepalived scenario, then they know the IP address.

File: deploy/tls/kustomization.yaml (hunk 7 above)
Comment (maelk) [reply deploy/tls/kustomization.yaml:13]: That's kustomize specific, to create a K8S secret with the certificate and key.

**Comment by dhellmann** [comment]
I would like to freeze go code changes for a few days to try to land https://github.com/metal3-io/baremetal-operator/pull/650 without having to rebase it, because rebasing will mean redoing the work from scratch.

/hold

File: ironic-deployment/keepalived-tls/kustomization.yaml (hunk 1 above)
Comment (s3rj1k) [reply ironic-deployment/keepalived-tls/kustomization.yaml]: @maelk "the mTLS part is not supported due to issue with PHA in go"
Can you please be more specific about "PHA in go"?
What this issues are?

Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml]: This is the exact issue :https://github.com/golang/go/issues/40521 Go does not support Post Handshake Authentication unfortunately, and our solution design was relying on PHA to separate the IPA endpoints that should not be authenticated from the rest : https://github.com/metal3-io/metal3-docs/blob/master/design/ironic_authentication.md

Comment (s3rj1k) [reply ironic-deployment/keepalived-tls/kustomization.yaml]: @maelk How about just using different DNS records that resolve to same IP?

Comment (maelk) [reply ironic-deployment/keepalived-tls/kustomization.yaml]: We are not using dns, but directly IP addresses. We cannot easily use different endpoints, so we are waiting for the ironic mTLS feature on IPA, that is WIP in Ironic.

**Comment by maelk** [comment]
We will address the change to the config folder, when we work on the metal3-dev-env adaptation. We currently do not have ways to test it in the CI otherwise.
